Added Redis Best practices recommended (#558)
elastisys/compliantkubernetes: docs/user-guide/additional-services/redis.md

Check out the [release notes](../../release-notes/redis.md) for the Redis cluster that runs in Compliant Kubernetes environments!

## Best Practices Recommended

> * **Eviction Policy**: Choose the eviction policy that works for your application.We highly recommend `volatile-lru` assuming you have expiring keys. If you are running something like a cache and don’t have an expiry set, you could consider `allkeys-lru`. Refer to [Key Eviction](https://redis.io/docs/reference/eviction/)

Reviewer: This should be and is set by us, and we by default set `allkeys-lru`.
So should we reconsider this option?

Author: Thanks Fredrik !! 

As once redis memory is full, an LRU algorithm kicks in, evicting least recently used keys. In `allkeys-lru`, it doesn't matter whether a key is expired or not and what is the TTL - the least used items will be evicted. 
In `volatile-lru` only expiring keys will be evicted using this algorithm.

Since we are recommending to set the `TTL`  as best practices , so yes we need to revisit the default eviction policy - I will open the mini-ADR and discuss in upcoming arch meeting.

Yes, this is set by us , but in order to make user more knowledgeable,  i would like to be available on public docs so user understand what kind of policy are available and what fits users the best to decide. I have re-phrased like below:

```suggestion
* **Eviction Policy**: Choose the [eviction policy](https://redis.io/docs/reference/eviction/) that works for your application. The default eviction policy for our Managed Redis is `allkeys-lru`, which means any key can be evicted under memory pressure irrespective of whether the key is expired or not. It will keep the most recently used keys and remove the least recently used (LRU) key.
```

Author: Opned mini-ADR - https://github.com/elastisys/ck8s-arch/issues/18

* **Key expiration**: Set an expiration value on your keys. An expiration removes keys proactively instead of waiting until there's memory pressure. When eviction happens because of memory pressure, it can cause more load on your server. For more information, see the documentation for the [EXPIRE](https://redis.io/commands/expire/) and [EXPIREAT](https://redis.io/commands/expireat/) commands.

> * **Set TTL**: If possible, take the advantage of expiring keys, such as temporary OAuth authentication keys. When you set the key, set it with the some expiration, and Redis will clean up for you. Refer to [TTL](https://redis.io/commands/ttl/)

Reviewer: What's the difference between these two? Isn't the `EXPIRE` command just the set (but this is also built in to the [`SET` command](https://redis.io/commands/set/)) and `TTL` is the get of this expiration time?

Author: yes Duplicate, removed.

* **Avoid expensive or blocking operations**: Since Redis command processing is single-threaded,  operations like the [KEYS](https://redis.io/commands/keys/) command, are expensive and should be avoided. You can avoid `KEYS` by using [SCAN](https://redis.io/commands/scan/) to reduce CPU spikes.
* **Monitor memory usage**: Monitor the usage in Grafana dashboard to ensure that you don't run out of memory and have the chance to scale your cache before seeing issues.

> * **Manage idle connection**: The number of connections to Redis increases if connections are not properly terminated. This can lead to bad performance. Therefore, we recommend to set `timeout` which allows you to set the number of seconds before idle client connections are automatically terminated.

Reviewer: The [`timeout`](https://github.com/redis/redis/blob/6.2/redis.conf#L119) setting seems to be a server setting so this is probably something we should set

Author: Yes this is a server side setting, but ideally we should set some timeout which is currently not set. 

```
127.0.0.1:6379> config get timeout
1) "timeout"
2) "0"
```
In order to make user more knowledgeable and the impact it brings , it should be recommended and accordingly user can decide what `timeout` values they want accordingly to their applications. I have re-phrased something like that below: 

```suggestion
* **Manage idle connection**: The number of connections to Redis increases if connections are not properly terminated. This can lead to bad performance. Therefore, we recommend to setting `timeout` which allows you to set the number of seconds before idle client connections are automatically terminated.
The default `timeout` for our Managed Redis is set to `0`, which means the idle clients do not timeout and remain connected until the client issues the termination..
```

Reviewer: I think I still would like it to be a bit more clear that this is something that they would need to tell us (send a ticket or tell at setup). Because this is something that they can set themselves until we've implemented [this issue](https://github.com/elastisys/compliantkubernetes-redis/issues/44) but it will not be persisted unless they tell us to set it for them.

But maybe adding something like `Since this is a server setting it can not be set by the user itself but needs to be set by the administrators. Please send a support ticket with the values you would like to set.`

Author: Done , updated.

Check out the [release notes](../../release-notes/redis.md) for the Redis cluster that runs in Compliant Kubernetes environments!

## Best Practices Recommended

> * **Eviction Policy**: Choose the eviction policy that works for your application.We highly recommend `volatile-lru` assuming you have expiring keys. If you’re running something like a cache and don’t have an expiry set, you could consider `allkeys-lru`. Refer- [Key Eviction](https://redis.io/docs/reference/eviction/)

Reviewer 2: ```suggestion
* **Eviction Policy**: Choose the eviction policy that works for your application. We highly recommend `volatile-lru` assuming you have expiring keys. If you’re running something like a cache and don’t have an expiry set, you could consider `allkeys-lru`. Refer to [Key Eviction](https://redis.io/docs/reference/eviction/)
```

* **Key expiration**: Set an expiration value on your keys. An expiration removes keys proactively instead of waiting until there's memory pressure. When eviction happens because of memory pressure, it can cause more load on your server. For more information, see the documentation for the [EXPIRE](https://redis.io/commands/expire/) and [EXPIREAT](https://redis.io/commands/expireat/) commands.

> * **Set TTL**: if possible , take the advantage of expiring keys ex-temporary OAuth authentication keys. When you set the key, set it with the some expiration, and Redis will clean up for you. Refer- [TTL](https://redis.io/commands/ttl/)

Reviewer 2: ```suggestion
* **Set TTL**: If possible, take the advantage of expiring keys, such as temporary OAuth authentication keys. When you set the key, set it with the some expiration, and Redis will clean up for you. Refer to [TTL](https://redis.io/commands/ttl/)
```

> * **Set TTL**: if possible , take the advantage of expiring keys ex-temporary OAuth authentication keys. When you set the key, set it with the some expiration, and Redis will clean up for you. Refer- [TTL](https://redis.io/commands/ttl/)

Reviewer 2: ```suggestion
* **Set TTL**: If possible, take the advantage of expiring keys, such as temporary OAuth authentication keys. When you set the key, set it with the some expiration, and Redis will clean up for you. Refer to [TTL](https://redis.io/commands/ttl/)
```

> * **Minimize memory fragmentation**: We recommend that you lower the `maxmemory` configuration for your instance to give yourself overhead in the case of high memory pressure. Refer- [Memory optimization](https://redis.io/docs/management/optimization/memory-optimization/)

Reviewer 2: Is this a recommendation for the user? Can you user do this? If not, then we should move it somewhere else.

Author: yeah agreed , i have removed it.

> * **Avoid expensive operations**: Since redis command processing is single-threaded,  operations like the [KEYS](https://redis.io/commands/keys/) command, are expensive and should be avoided. You can avoid KEYS by using [SCAN](https://redis.io/commands/scan/) to reduce CPU spikes.

Reviewer 2: ```suggestion
* **Avoid expensive or blocking operations**: Since Redis command processing is single-threaded, operations like the [KEYS](https://redis.io/commands/keys/) command are expensive and should be avoided. You can avoid KEYS by using [SCAN](https://redis.io/commands/scan/) to reduce CPU spikes.
```

> * **Monitor memory usage**: Monitor the usage in grafana dashboard to ensure that you don't run out of memory and have the chance to scale your cache before seeing issues.

Reviewer 2: ```suggestion
* **Monitor memory usage**: Monitor the usage in Grafana dashboard to ensure that you don't run out of memory and have the chance to scale your cache before seeing issues.
```

> * **Manage idle connection**: Number of connections to redis increases if connections are not properly terminated and can have bad performance. so we recommend , to set `timeout` which allows you to set the number of seconds before idle client connections are automatically terminated.

Reviewer 2: Is this recommendation something the user can do?

```suggestion
* **Manage idle connection**: The number of connections to Redis increases if connections are not properly terminated. This can lead to bad performance. Therefore, we recommend  to set `timeout` which allows you to set the number of seconds before idle client connections are automatically terminated.
```
